### PR TITLE
SI-7364 Allow raw types in parent position in Java sources

### PR DESCRIPTION
**Pull request by retronym**
To make this work, this commit simply restricts parent
type argument inference to Scala source files.

The surrounding code has also been refactored to avoid a var.

Review by @xeno-by (last seen in this area in 40063b00)

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 1]
```diff
@@ -1571,9 +1571,9 @@ trait Typers extends Adaptations with Tags {
         }
         typedType(decodedtpt)
       } else {
-        var supertpt = typedTypeConstructor(decodedtpt)
+        val supertpt = typedTypeConstructor(decodedtpt)
```
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:1574]: Yep, I've overlooked this one. Thank you!

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 2]
```diff
@@ -1581,14 +1581,17 @@ trait Typers extends Adaptations with Tags {
             ctor setType supertpe // this is an essential hack, otherwise it will occasionally fail to typecheck
             atPos(supertpt.pos.focus)(supercall)
           } match {
-            case EmptyTree => MissingTypeArgumentsParentTpeError(supertpt)
-            case tpt => supertpt = TypeTree(tpt.tpe) setPos supertpt.pos  // SI-7224: don't .focus positions of the TypeTree of a parent that exists in source
+            case EmptyTree => MissingTypeArgumentsParentTpeError(supertpt); supertpt
```
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:1584]: Isn't it common for `...Error` methods to mark an argument tree as erroneous and throw it back to the caller?

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 3]
```diff
@@ -1581,14 +1581,17 @@ trait Typers extends Adaptations with Tags {
...
+            case tpt       => TypeTree(tpt.tpe) setPos supertpt.pos  // SI-7224: don't .focus positions of the TypeTree of a parent that exists in source
           }
         }
+
+        val supertptInferred = if (supertparams.isEmpty || context.unit.isJava) supertpt else inferParentTypeArgs
```
Comment (xeno-by) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: Maybe `supertptWithTargs`?

**Comment by soc** [comment]
Might this help with the issues regarding Swing, too?

**Comment by retronym** [comment]
Nope, this is just for .java source files fed into Scalac.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 2 above)
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:1584]: Yes, but I'm a bit reluctant to change this case without knowing the intentions of the orignal author.

**Comment by retronym** [comment]
@xeno-by I've changed that identifier, but left the error tree handling as it was out of conservatism.

**Comment by gkossakowski** [comment]
@xeno-by: ping.

**Comment by xeno-by** [comment]
lgtm

**Pull request merged**